### PR TITLE
Add exercises for header units and modules

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This PR adds exercises for header units and modules. Since there is no cmake support for both features yet, I could only add Makefiles for g++.

- [x] Merge #200 first
- [x] Merge #222 first

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
I especially need someone to read through the Makefiles. This is not my domain :)

File: code/header_units/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+## Instructions
+
+* convert the `#include "Complex.hpp"` to a header unit import
+  * you will need to edit the Makefile and add a new target to precompile `Complex.hpp`
+* convert the standard library includes in `randomize.cpp` to header unit imports
+  * you will need to precompile the standard library headers in your Makefile
+  * with g++-11 and g++12 you cannot yet precompile `<iostream>` and `<ostream>`, so keep those as a regular includes
```
Comment (hageboeck) [inline code/header_units/README.md:8]: `g++-11` or `g++11`?

File: code/header_units/solution/Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	${CXX} -std=c++20 -fmodules-ts -x c++-system-header $@
+
+numeric:
+	${CXX} -std=c++20 -fmodules-ts -x c++-system-header $@
+
+random:
+	${CXX} -std=c++20 -fmodules-ts -x c++-system-header $@
+
+vector:
+	${CXX} -std=c++20 -fmodules-ts -x c++-system-header $@
```
Comment (hageboeck) [inline code/header_units/solution/Makefile:16]: Do you want to unify this to one target?

Comment (bernhardmgruber) [reply code/header_units/solution/Makefile:16]: We could, but these can be built in parallel. Can I express this in one target?

File: code/header_units/README.md (hunk 1 above)
Comment (bernhardmgruber) [reply code/header_units/README.md]: Thx!

File: code/header_units/solution/Makefile (hunk 2 above)
Comment (sponce) [reply code/header_units/solution/Makefile:16]: I do not see why saying 
```Makefile
algorithm numeric random vector:
	${CXX} -std=c++20 -fmodules-ts -x c++-system-header $@
```
would not be strictly identical to the current code, and in particular would not allow to run in parallel. To be checked though, but it looks like pure syntax.

**Pull request merged**